File: Python/4_Merge Sort/Merge Sort.py
def merge(left, right):

    result = []
    i, j = 0, 0
    
    while (len(result) < len(left) + len(right)):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
        
        if i == len(left) or j == len(right):
            result.extend(left[i:] or right[j:])
            break
        
    return result

def merge_sort(l):
    if len(l) < 2:
        return l
    
    mid = len(l)//2
    left = merge_sort(l[:mid])
    right = merge_sort(l[mid:])
    
    return merge(left, right)

a = [3,4,5,1,2,8,3,7,6]
print(merge_sort(a))

File: Javascript/4_Merge Sort/Merge Sort.js
function merge(left, right) {
    let result = [], i = 0, j = 0;

    while(i < left.length && j < right.length){
        if(left[i] < right[j]){
            result.push(left[i]);
            i++;
        }else{
            result.push(right[j]);
            j++;
        }
    }

    return result.concat(left.slice(i)).concat(right.slice(j));
}

function mergeSort(array) {
    if(array.length < 2){
        return array;
    }

    const mid = Math.floor(array.length / 2);
    const left = mergeSort(array.slice(0, mid));
    const right = mergeSort(array.slice(mid));

    return merge(mergeSort(left), mergeSort(right));
}

array = [3, 4, 5, 1, 2, 8, 3, 7, 6]
console.log(mergeSort(array));